File: src/test/java/com/letgo/book/unit/application/SpyDomainEventSubscriber.java
package com.letgo.book.unit.application;

import com.letgo.shared.application.event.DomainEventSubscriber;
import com.letgo.shared.domain.DomainEvent;

import static org.junit.jupiter.api.Assertions.*;

final public class SpyDomainEventSubscriber implements DomainEventSubscriber {
    private boolean hasBeenCalled = false;
    private final DomainEvent expectedEvent;

    public SpyDomainEventSubscriber() {
        this.expectedEvent = null;
    }

    public SpyDomainEventSubscriber(DomainEvent expectedEvent) {
        this.expectedEvent = expectedEvent;
    }

    @Override
    public boolean isSubscribedTo(DomainEvent event) {
        return true;
    }

    @Override
    public void consume(DomainEvent event) {
        hasBeenCalled = true;
        assertNotNull(expectedEvent);
        assertEquals(expectedEvent.aggregateId(), event.aggregateId());
        assertEquals(expectedEvent.name(), event.name());
        assertEquals(expectedEvent.body(), event.body());
        assertEquals(expectedEvent.occurredOn(), event.occurredOn());
    }

    public boolean hasBeenCalled() {
        return hasBeenCalled;
    }
}

File: src/test/java/com/letgo/shared/infrastructure/bus/event/SpyDomainEventSubscriber.kt
package com.letgo.shared.infrastructure.bus.event

import com.letgo.shared.application.bus.event.DomainEventSubscriber
import com.letgo.shared.domain.DomainEvent
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull

class SpyDomainEventSubscriber(
    private val expectedEvent: DomainEvent? = null
) : DomainEventSubscriber {
    private var hasBeenCalled = false

    override fun isSubscribedTo(event: DomainEvent): Boolean = true

    override fun consume(event: DomainEvent) {
        hasBeenCalled = true
        expectedEvent.let {
            assertNotNull(it)
            assertEquals(it!!.aggregateId(), event.aggregateId())
            assertEquals(it.name(), event.name())
            assertEquals(it.body(), event.body())
            assertEquals(it.occurredOn(), event.occurredOn())
        }
    }

    fun hasBeenCalled(): Boolean = hasBeenCalled
}
